### DISCUSSION TITLE
simple auto-node

### DISCUSSION THREAD

**Discussion by Spaced-out-games** [leon-thomm/Ryven#108] [Ideas]
I'm working on making a node lib for this (mostly integrating already-existing libraries into Ryven) and I don't feel like defining each of them by hand, so I was thinking that maybe you could add a simple-auto-node method

The simplest definition of a node is as follows:

```python
class SimpleNode(Node):
	"""docstring"""
	title = "title"

	tags = ["str"]

	init_inputs = [
		NodeInputBP(dtype=dtypes.Data(default=1))
		]
	init_outputs = [NodeOutputBP()]
	color ="#HEXCOL"
	def update_event(self, inp=-1):
		input = self.input
		result = input
		self.set_output_val(index,result)

```

I started making an algorithm that takes in a function and outputs something like this, but I never finished it. Plus,such an algorithm would make prototyping much, much quicker.

**Comment by leon-thomm** [comment]
I want to make the general remark that often when you want to automatically parse functions into nodes, you are already at a level of detail that should not be exposed to this level of visualization. However, there are still many cases where this can be useful, but it should rather be seen as a starting point to generate a template code that can then be optimized further, instead of a complete start-to-finish workflow. This should be easy to write, for some inspiration something like the following should get you starting quickly:
```python
import inspect

def node_from_function(name: str, call: str, obj: object, color: str):
    """Create a node class definition from a given function.

    Parameters
    ----------
    name: str
        Becomes the title of the node.
    call: str
        Represents the full function call without parentheses, e.g. `math.sqrt`
    obj: object
        The function object. Its souce will be parsed using the inspect module,
        which might not work for some functions, expecially those of C bindings.
    color: str
        The color hex code the node will get.

    Returns
    -------
    node_def: str
        The code defining the node class.
    """
    
    try:
        sig = inspect.getfullargspec(obj)
    except Exception as e:
        raise Exception(f"Could not parse source of {name}.")
    
    inputs = '\n'.join([f"NodeInputBP('{param_name}')," for param_name in sig.args])
    node_name = f'{name}_Node'

    node_def = f"""
class {node_name}(Node):
    \"\"\"{obj.__doc__}\"\"\"

    title = '{name}'
    init_inputs = [
        {inputs}
    ]
    init_outputs = [
        NodeOutputBP(),
    ]
    color = '{color}'

    def update_event(self, inp=-1):
        self.set_output_val(0, {call}({
            ', '.join([f'self.input({i})' for i in range(len(sig.args))]) 
                                        }))
"""

    return node_def
```
E.g. `node_from_function('sqrt', 'math.sqrt', math.sqrt, '#aabbcc')` would give you
```python
class sqrt_Node(Node):
    """Return the square root of x."""

    title = 'sqrt'
    init_inputs = [
        NodeInputBP('x'),
    ]
    init_outputs = [
        NodeOutputBP(),
    ]
    color = '#aabbcc'

    def update_event(self, inp=-1):
        self.set_output_val(0, math.sqrt(self.input(0)))
```
This code uses `inspect` to get the arguments of the function, which only works for functions that can be fully inspected at Python level; e.g. `math.hypot` cannot be parsed this way. Feel free to improve it and post it here, I may update the code then.

**Reply by leon-thomm** [reply]
next step might be adding `dtype` widgets to inputs which are annotated to have a "primitive" type
```python
inputs = '\n'.join([
    f"""NodeInputBP('{param_name}'{
        f", dtype=dtypes.Data(size='s')"
        if sig.annotations[param_name] in (int, float, bool, str, ...)
        else ''
    }),"""
    for param_name in sig.args
])
```
though `...` should be expanded based on [`typing`](https://docs.python.org/3/library/typing.html) type hints which I am not too familiar with.

**Reply by Spaced-out-games** [reply]
> I want to make the general remark that often when you want to automatically parse functions into nodes, you are already at a level of detail that should not be exposed to this level of visualization. However, there are still many cases where this can be useful, but it should rather be seen as a starting point to generate a template code that can then be optimized further, instead of a complete start-to-finish workflow. This should be easy to write, for some inspiration something like the following should get you starting quickly:

I completely agree that it should only be used to generate a simple outline , but in some cases, the code generated is good enough to use. Of course, testing each node would be a requirement, though.
I have a question, though...Does Ryven support other object types (or at least the implementation of them), or is it primitives only? I'm assuming it does and making sure.

**Reply by leon-thomm** [reply]
nonono, Ryven doesn't care at all about the type of data passed between the nodes, that's up to the nodes developer, I just referenced primitives above because it doesn't make a lot of sense to have a `dtype` input widget for complex object types, but you can also just give them all dtypes at first, of course

**Reply by Spaced-out-games** [reply]
I originally was going to make something similar to Ryven, but I eventually found it and have essentially ditched it since you already have something working with a presentable UI. (see https://github.com/Spaced-out-games/BrainBoard) Considering Ryven is a visual programming interface for Python, and Python is a staple of machine learning, I've been working on integrating NumPy, PyTorch, and eventually requests into Ryven. Not to mention, the visual nature of Ryven makes it ideal for visualizing data structures on-the-fly.

**Reply by leon-thomm** [reply]
awesome! I can't wait to see progress there. The biggest part Ryven is missing is various large and flexible node packages.

**Reply by Spaced-out-games** [reply]
Yeah, that's why auto-node needs implemented! (assuming one of us can write a good algorithm)

I'll try to take your code and improve upon it, and assuming it works better, post it here.

One thing to add would be automatic tag generation since tags are pretty predictable,
 numpy.array(), tags could be ["numpy","array","ndarray"] = [class, method, output type]

**Reply by Spaced-out-games** [reply]
@leon-thomm I haven't checked if you already have one, but maybe you should set up a repo where node packages (such as the ones I'm going to make) can go?
Also, by the way, when running ryven.run, ryven fails to load (I found a fix: `export QT_MAC_WANTS_LAYER=1 `in console)

file structure:
```
Ryven-packages repo
	packages
		torch	<--optional package
			widgets.py
			nodes.py
			misc files(README, LICENSE...)
		numpy	<--optional package
			widgets.py
			nodes.py
			misc files(README, LICENSE...)
	auto-import.py	<--goes into packages and searches for folders and imports each nodes.py that the user wants. Every user downloads this.
```

**Reply by Spaced-out-games** [reply]
crap, the tabs didn't work

**Reply by leon-thomm** [reply]
> crap, the tabs didn't work

fixed it :)

Currently, there's no such repo, although I really really want this. There just hasn't been enough content so far; most node packages created so far were either closed source or designed for one specific use. The only existing large and somewhat maintained package that is designed for general use is the [PythonOCC package](https://github.com/Tanneguydv/Pythonocc-nodes-for-Ryven). Though, if you are creating these large general-purpose packages, it probably makes sense that I do this soon...

**Reply by Spaced-out-games** [reply]
From what I understand, Numpy is fair game, given the license is included

Source: https://numpy.org/doc/stable/license.html

Also, the package I'd write would not be distributing the source of NumPy, just implementation of it (my package is useless unless the user has NumPy installed), but still better safe than sorry, ill provide the license in the project folder

**Reply by leon-thomm** [reply]
Ah, don't worry if you're just using the `numpy` python library, that is not "redistribution of source code". You can (and should) choose the license that you want to put on *your* code.

**Reply by Spaced-out-games** [reply]
I'm honestly quite new to licensing things that I make, since all of them are amateur and my own projects, just to learn the ropes  I'm thinking open-source w/credit is my way to go since the entire intent is to make machine learning accessible to everyone, using your framework. it _should_ be pretty straightforward, generating code, testing them, and trying to fix whatever bugs I come across. Plus, since Ryven isn't picky about datatypes, there's no need to define the pins of the nodes as certain types. Thank goodness

**Comment by Spaced-out-games** [comment]
I'll probably hand-code a visualize() node that draws a graphical representation of data-driven objects such as numpy.ndarrays, or torch.tensor

**Reply by Spaced-out-games** [reply]
hey, I found a workaround for nodes like math.hypot...We might not be able to find what arguments are in a function uninspectable, but we can find out _how many_ are there. We can use this function(I didn't write this function) to find that:
```
def num_args(f):
  if isfunction(f):
    return len(getargspec(f).args)
  else:
    spec = f.__doc__.split('\n')[0]
    args = spec[spec.find('(')+1:spec.find(')')]
    return args.count(',')+1 if args else 0
```
Using this, we can name each parameter a - z, depending on the result of num_args.
I implemented it into your node_from_function, and instead of throwing an error on an exception, it just uses num_args as a fallback option and uses that. It also adds a warning to the top of the code it generates. Check out automation.py on my [repo](https://github.com/Spaced-out-games/ryven-pytorch-lib).
Math.hypot:
```
class MathHypotNode_Node(Node):
    """Return the Euclidean distance, sqrt(x*x + y*y)."""

    title = 'MathHypotNode'
    init_inputs = [
        NodeInputBP(dtype=dtypes.Data(default=1), label = 'x'),
NodeInputBP(dtype=dtypes.Data(default=1), label = 'y'),
    ]
    init_outputs = [
        NodeOutputBP(),
    ]
    color = '#aaaaaa'

    def update_event(self, inp=-1):
        self.set_output_val(0, hypot(self.input(0), self.input(1)))
```

for some reason, math.hypot did not add the warning comment. it should have
```
"""
WARNING: Module MathHypotNode_Node was generated using the fallback option. May contain bugs
"""
``` 
at the top of the definition.

**Reply by Spaced-out-games** [reply]
@leon-thomm I have a file in my repo and it has so far pretty much been plug'n play, minus node exports, so the code gen needs refining. The entire math module has been converted (minus builtin attributes like __ add __) and needs testing. It's under package in my ryven-pytorch-lib repo, just download all three .py files

**Reply by leon-thomm** [reply]
I don't really understand how this should be working. While relying on the docstrings is probably not so rigurous, if I understand the code correctly it simply looks for the first parentesis in the first line of the docstring of the function and counts the `,` to get the number of arguments. This already breaks when having signatures like `foo(*args)`, which is the case for many, if not almost all such binding functions; in fact this is even the case for `math.hypot`
```python
def hypot(*coordinates): # real signature unknown; restored from __doc__
    """
    hypot(*coordinates) -> value
    
    Multidimensional Euclidean distance from the origin to a point.
    [...]
```
And indeed, if I run `num_args(math.hypot)` I get 1, not 2. So, I think this also explains why it didn't generate the string, because your code did not use `num_args`, but did something else... :thinking:

**Reply by Spaced-out-games** [reply]
Maybe pay attention to that star? Idk.
If I’m not mistaken, f(*(0,0)) is equal to f(0,0)
That explains why it works off of one argument. Because the argument itself is dynamic in length, you can pass anything into it and it works, mostly because math.hypot is multi-dimensional math

**Reply by Spaced-out-games** [reply]
Here is an example:
```
point = (0,0,0)#tuple (0,0,0)
print(point)
#prints (0,0,0)
print(*point)#0 0 0
#prints 0,0,0
print(0,0,0)
#prints 0,0,0

#so shouldn't *(0,0,0) equal 0,0,0?    <--(   notice the lack of parentheses around 0,0,0   )
```

**Reply by Spaced-out-games** [reply]
I'm starting to think that functions that fail to give arguments should be automatically given 1 parameter, so instead of
```
class MathHypotNode_Node(Node):
    """Return the Euclidean distance, sqrt(x*x + y*y)."""

    title = 'MathHypotNode'
    init_inputs = [
        NodeInputBP(dtype=dtypes.Data(default=1), label = 'x'),
NodeInputBP(dtype=dtypes.Data(default=1), label = 'y'),
    ]
    init_outputs = [
        NodeOutputBP(),
    ]
    color = '#aaaaaa'

    def update_event(self, inp=-1):
        self.set_output_val(0, hypot(self.input(0), self.input(1)))
```

we do

```
class MathHypotNode_Node(Node):
    """Return the Euclidean distance, sqrt(x*x + y*y)."""

    title = 'MathHypotNode'
    init_inputs = [
        NodeInputBP(dtype=dtypes.Data(default=1), label = 'InputTuple'), #edits here
    ]
    init_outputs = [
        NodeOutputBP(),
    ]
    color = '#aaaaaa'

    def update_event(self, inp=-1):
        self.set_output_val(0, hypot(*self.input))#and here.
        #Honestly, it should be this way from the beginning, instead of having f(self.input(0)...self.input(n), just do this
```

**Reply by Spaced-out-games** [reply]
alright, I have a demo torch module ready for testing, but Ryven imports the nodes pretty slowly. I had to comment out the entire file, minus five or so nodes to test it. I know that Ryven relies on crowdsourcing, so it's probably a bad idea to ask you to look into it, right?

however, if you are up to the task, I'd suggest imports to be a background task, so that people aren't waiting several minutes for a module to import. Either that, or an improvement on import speed...

**Reply by Spaced-out-games** [reply]
@leon-thomm See above. Also, am I bothering you with all these posts? Maybe I'm talking too much

**Reply by leon-thomm** [reply]
No don't worry, we all need a place to learn and brainstorm on concepts we are not really familiar with just yet. GitHub Discussions is finally a good place on github to do that, much better than e.g. issues. It might just take me some time occasionally to answer when there's too much to do, apologies in advance. I do read everything eventually.

The `*args` syntax is not what you think it is, it's pythons way to denote an arbitrary number of arguments. For example
```python
def foo(*args):
    for arg in args:
        print(arg)
```
takes any list of arguments and prints them. For example `foo(1, 2, 'asdf', 42.1, False, ('t','u','p','l','e'))` is totally valid. The `*` operator is used because it also relates to *tuple unpacking* (the list of arguments is just a big tuple, in the case above we can refer to this tuple through `args`). An analogous principle applies to keyword arguments, denoted as `foo(**kwargs)`. This should give you enough keywords to quickly google for the remaining details.

Therefore, for functions with a variable number of inputs, one could argue to also make the node have a dynamic number of inputs as well. This, however, introduces state; your node becomes stateful, which itself is supported and necessary for all kinds of sophisticated nodes, but it heavily complicates development and introduces a huge domain of possible hard to find bugs, so yes we might just give it 1 input and we can still make an additional node `Gather` which simply bundles multiple arguments for a function and nothing else (so it is simple and easy to debug).

Regarding speed, this is most likely not a Ryven issue. Importing nodes is as fast as your node package is. I used large packages which loaded really fast, so if it takes ages there might be an issue with your code, or your dependencies are just very large (which I can't do anything about)

**Reply by Spaced-out-games** [reply]
ah ok. When it comes to tuple unpacking, its something I already knew, but just couldn't spit it out right, lol. Pytorch is definitely a gargantuan-sized package (there are more than a thousand methods and objects, which themselves have methods and objects inside)

Maybe I can separate them into sub-packages? Commonly used methods like torch.from_numpy and torch.Tensor could be in default.py, while the lesser-used methods can be in a file called extras.py. It would make loading less of an issue.

Now that I think about it, maybe the node importer can be asynchronous because the problem is that Ryven is unresponsive while importing.

One thing Ryven is lacking is built-in support(discussion #109). I could implement those datatypes relatively quickly, assuming there aren't 1000 methods on str() lol. Another is that nodes are not able to be deleted. All nodes are given four options automatically, why not add a fifth?

anyway, I'd better get back to schoolwork; ill work on this all later

**Reply by leon-thomm** [reply]
While there is currently no way to define actual hierarchies of packages (of course you can distribute the sources into different files and python packages as you like, but they still have to be imported when the nodes package is imported), you can

- simply define different packages (brings the advantage of pushing you to make the packages modular and independent), or
- import dynamically

For the second approach you can really go crazy by distributing the different PyTorch-subpackage-specific nodes into different respective python modules and write them such that the pytorch sub-package is only imported when any of the nodes is instantiated, not before, and then you have again one single ryven nodes package but pytorch sources are imported on demand; but that might be overkill.

> Now that I think about it, maybe the node importer can be asynchronous because the problem is that Ryven is unresponsive while importing.

Hm, yes that's a good point, though I'm not sure how difficult this would be.

**Reply by Spaced-out-games** [reply]
asyncio...is honestly horrifying

I wish you could make a file that is a self-contained thread and then asyncio. run that file. It would honestly make developing with it easier

**Comment by Spaced-out-games** [comment]
@leon-thomm
Alright, so I've got a stable Ryven auto-node up and running. It can convert functions into Ryven - ready nodes. You can find it here:
https://github.com/Spaced-out-games/Ryven-Package-Manager/blob/9c84b6dd2514390e2c783b42d535221a88357765/tools/code_gen.py

It should theoretically work for every function(even non-python ones), but the (non-python) nodes will have nameless arguments. 
It only works for individual functions and methods at the moment, and the plan is to add the ability to nodify an entire module, recursively. I tested it on the `math` module and it worked pretty well - Just needed a few fixes(now implemented into the function natively) and some import statements on the top and bottom and it's ready to go!
